src/lustre/lustreTypeChecker.ml

```diff
@@ -0,0 +1,1379 @@
+             then 
+             R.ok (
+                 List.fold_left union (add_ty_decl ctx ename)
+                   ((List.map ((Lib.flip singleton_ty)
+                                 (LA.UserType (pos, ename))) econsts)
+                    @ Lib.list_apply ((List.map2 (Lib.flip singleton_const)
+                         (List.map (fun i -> LA.Ident (pos, i)) econsts) econsts)) (LA.UserType (pos, ename)))
+               )
+             else
+               type_error pos "Cannot redeclare constants or enums")
+      | _ -> R.ok (add_ty_syn ctx i ty))
+  | LA.FreeType (_, i) -> R.ok (add_ty_decl ctx i)
+
+and tc_ctx_of_node_decl: Lib.position -> tc_context -> LA.node_decl -> tc_context tc_result
+  = fun pos ctx (nname, _, _ , ip, op,_ ,_ ,_) ->
+  Log.log L_debug
+    "Extracting typing context from node declaration: %a"
+    LA.pp_print_ident nname
+  ; if (member_ty ctx nname)
+    then type_error pos ("Duplicate node detected with name: " ^ nname)
+    else build_node_fun_ty pos ctx ip op
+         >>= fun fun_ty -> R.ok (add_ty ctx nname fun_ty)
+(** computes the type signature of node or a function *)
+
+and tc_ctx_of_contract_node_decl: Lib.position -> tc_context
+                                  -> LA.contract_node_decl
+                                  -> tc_context tc_result
+  = fun pos ctx (cname, params, inputs, outputs, contrat) ->
+  Log.log L_debug
+    "Extracting typing context from contract declaration: %a"
+    LA.pp_print_ident cname
+  ; if (member_ty ctx cname)
+    then type_error pos ("Duplicate node detected with name: " ^ cname)
+    else build_node_fun_ty pos ctx inputs outputs
+         >>= fun fun_ty -> R.ok (add_ty ctx cname fun_ty)
+
+and tc_context_of: tc_context -> LA.t -> tc_context tc_result
+  = fun ctx ->
+  let rec tc_context_of': tc_context -> LA.declaration -> tc_context tc_result
+    = fun ctx' ->
+    function
+    | LA.ConstDecl (_, const_decl) -> tc_ctx_const_decl ctx' const_decl
+    | LA.NodeDecl (pos, node_decl) -> tc_ctx_of_node_decl pos ctx' node_decl
+    | LA.FuncDecl (pos, node_decl) -> tc_ctx_of_node_decl pos ctx' node_decl
+    | LA.ContractNodeDecl (pos, contract_decl) ->
+       tc_ctx_of_contract_node_decl pos ctx' contract_decl
+    | _ -> R.ok ctx'
+  in
+  function
+  | [] -> R.ok ctx
+  | d :: tl ->
+     tc_context_of' ctx d
+     >>= fun ctx' -> tc_context_of (union ctx' ctx) tl
+     >>= fun c -> R.ok c
+(** Obtain a global typing context, get constants and function decls*)
+
+and build_type_context: tc_context -> LA.t -> tc_context tc_result
+  = fun ctx ->
+  function
+  | [] -> R.ok ctx
+  | TypeDecl (_, ty_decl) :: rest ->
+     tc_ctx_of_ty_decl ctx ty_decl
+     >>= fun ctx' -> build_type_context ctx' rest
+  | _ :: rest -> build_type_context ctx rest
+(** Process top level type declarations and make a type context with 
+ * user types, enums populated *)
+               
+and is_type_well_formed: tc_context -> tc_type -> bool
```

Comment:
I suggest to change the return type of this function (and maybe its name accordingly) to a result type that stores an error message in case of`Error` explaining why it is ill-formed. In particular, it should return a message saying `"Unknown type " ^ i` when it cannot find a UserType identifier (review test `undeclared_type_01.lus`)

Reply:
using `check_type_well_formed: tc_context -> tc_type -> unit tc_result` now.